File: backend/app/api/v1/endpoints/appointments.py
from typing import List, Any
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy.orm import Session
from datetime import datetime

from app.api import deps
from app.models.models import Appointment, User, Business
from app.schemas.appointment import AppointmentCreate, AppointmentUpdate, Appointment as AppointmentSchema
from app.services.audit_service import create_audit_log

router = APIRouter()

@router.get("/business/{business_id}", response_model=List[AppointmentSchema])
def read_appointments(
    business_id: int,
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    db: Session = Depends(deps.get_db),
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Retrieve appointments for a business.
    """
    if current_user.role != "admin":
         business = db.query(Business).filter(Business.id == business_id).first()
         if not business or business.user_id != current_user.id:
             raise HTTPException(status_code=400, detail="Not enough permissions")

    appointments = (
        db.query(Appointment)
        .filter(Appointment.business_id == business_id)
        .order_by(Appointment.appointment_time.desc())
        .offset(skip)
        .limit(limit)
        .all()
    )
    return appointments

@router.post("/", response_model=AppointmentSchema)
def create_appointment(
    *,
    request: Request,
    db: Session = Depends(deps.get_db),
    appointment_in: AppointmentCreate,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Create new appointment.
    """
    if current_user.role != "admin":
         business = db.query(Business).filter(Business.id == appointment_in.business_id).first()
         if not business or business.user_id != current_user.id:
             raise HTTPException(status_code=400, detail="Not enough permissions")

    appointment = Appointment(**appointment_in.model_dump())
    db.add(appointment)
    create_audit_log(
        db,
        user_id=current_user.id,
        business_id=appointment_in.business_id,
        operation="appointment.create",
        resource_type="appointment",
        new_values=appointment_in.model_dump(mode="json"),
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
    )
    db.commit()
    db.refresh(appointment)
    return appointment

@router.put("/{id}", response_model=AppointmentSchema)
def update_appointment(
    *,
    request: Request,
    db: Session = Depends(deps.get_db),
    id: int,
    appointment_in: AppointmentUpdate,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Update an appointment.
    """
    appointment = db.query(Appointment).filter(Appointment.id == id).first()
    if not appointment:
        raise HTTPException(status_code=404, detail="Appointment not found")

    if current_user.role != "admin":
         business = db.query(Business).filter(Business.id == appointment.business_id).first()
         if not business or business.user_id != current_user.id:
             raise HTTPException(status_code=400, detail="Not enough permissions")

    old_values = {
        "customer_name": appointment.customer_name,
        "customer_phone": appointment.customer_phone,
        "status": appointment.status,
    }
    for field, value in appointment_in.model_dump(exclude_unset=True).items():
        setattr(appointment, field, value)

    create_audit_log(
        db,
        user_id=current_user.id,
        business_id=appointment.business_id,
        operation="appointment.update",
        resource_type="appointment",
        resource_id=id,
        old_values=old_values,
        new_values=appointment_in.model_dump(exclude_unset=True, mode="json"),
        ip_address=request.client.host if request.client else None,
        user_agent=request.headers.get("user-agent"),
    )
    db.add(appointment)
    db.commit()
    db.refresh(appointment)
    return appointment

File: backend_legacy_nodejs/routes/appointments.js
const express = require('express');
const router = express.Router();
const db = require('../database');

// Get all appointments for a business
router.get('/business/:businessId', async (req, res) => {
  try {
    const { businessId } = req.params;
    const { rows } = await db.query('SELECT * FROM appointments WHERE business_id = $1', [businessId]);
    res.json(rows);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// Get a single appointment by id
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { rows } = await db.query('SELECT * FROM appointments WHERE id = $1', [id]);
    if (rows.length === 0) {
      return res.status(404).json({ error: 'Appointment not found' });
    }
    res.json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// Create a new appointment
router.post('/', async (req, res) => {
  try {
    const { business_id, customer_name, customer_phone, appointment_time, service_type } = req.body;
    const { rows } = await db.query(
      'INSERT INTO appointments (business_id, customer_name, customer_phone, appointment_time, service_type) VALUES ($1, $2, $3, $4, $5) RETURNING *',
      [business_id, customer_name, customer_phone, appointment_time, service_type]
    );
    res.status(201).json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// Update an appointment
router.put('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { customer_name, customer_phone, appointment_time, service_type, status } = req.body;
    const { rows } = await db.query(
      'UPDATE appointments SET customer_name = $1, customer_phone = $2, appointment_time = $3, service_type = $4, status = $5 WHERE id = $6 RETURNING *',
      [customer_name, customer_phone, appointment_time, service_type, status, id]
    );
    if (rows.length === 0) {
      return res.status(404).json({ error: 'Appointment not found' });
    }
    res.json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// Delete an appointment
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { rowCount } = await db.query('DELETE FROM appointments WHERE id = $1', [id]);
    if (rowCount === 0) {
      return res.status(404).json({ error: 'Appointment not found' });
    }
    res.status(204).send();
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

module.exports = router;
